ultraplot/internals/rcsetup.py

```diff
@@ -587,22 +614,33 @@ def __iter__(self):
         yield from sorted(dict.__iter__(self))
 
     def __getitem__(self, key):
-        key, _ = self._check_key(key)
-        return dict.__getitem__(self, key)
+        with self._lock:
+            key, _ = self._check_key(key)
+            # Check thread-local storage first
+            if key in self._local.changes:
+                return self._local.changes[key]
+            # Check global dictionary (will raise KeyError if not found)
+            return dict.__getitem__(self, key)
 
     def __setitem__(self, key, value):
-        key, value = self._check_key(key, value)
-        if key not in self._validate:
-            raise KeyError(f"Invalid rc key {key!r}.")
-        try:
-            value = self._validate[key](value)
-        except (ValueError, TypeError) as error:
-            raise ValueError(f"Key {key}: {error}") from None
-        if key is not None:
-            dict.__setitem__(self, key, value)
-
-    @staticmethod
-    def _check_key(key, value=None):
+        with self._lock:
+            key, value = self._check_key(key, value)
+            # Validate the value
+            try:
+                value = self._validate[key](value)
+            except KeyError:
+                # If key doesn't exist in validation, add it with default validator
+                self._validate[key] = lambda x: x
+                # Re-validate with new validator
+                value = self._validate[key](value)
+            except (ValueError, TypeError) as error:
+                raise ValueError(f"Key {key}: {error}") from None
+            if key is not None:
+                # Store in both thread-local storage and main dictionary
+                self._local.changes[key] = value
+                dict.__setitem__(self, key, value)
```

Comment:
This block doesn't make sense to me. I think we need to test if we're in the context manager block and then only make a local change if we are. Otherwise threads will make changes to the global settings which could interfere with each other.

Reply:
I am leaning towards moving this logic (thread safety) inside the `Configurator` we both need to control ultraplot's rc and matplotlib's rc making this PR a bit moot and #384 an essential step in this direction.